ackhandler/sent_packet_handler.go

```diff
@@ -211,59 +204,124 @@ func (h *sentPacketHandler) ReceivedAck(ackFrame *frames.AckFrame, withPacketNum
 
 			if packetNumber >= ackRange.FirstPacketNumber { // packet i contained in ACK range
 				if packetNumber > ackRange.LastPacketNumber {
-					return fmt.Errorf("BUG: ackhandler would have acked wrong packet 0x%x, while evaluating range 0x%x -> 0x%x", packetNumber, ackRange.FirstPacketNumber, ackRange.LastPacketNumber)
-				}
-				h.ackPacket(el)
-				ackedPackets = append(ackedPackets, congestion.PacketInfo{Number: packetNumber, Length: packet.Length})
-			} else {
-				retransmitted := h.nackPacket(el)
-				if retransmitted {
-					lostPackets = append(lostPackets, congestion.PacketInfo{Number: packetNumber, Length: packet.Length})
+					return nil, fmt.Errorf("BUG: ackhandler would have acked wrong packet 0x%x, while evaluating range 0x%x -> 0x%x", packetNumber, ackRange.FirstPacketNumber, ackRange.LastPacketNumber)
 				}
+				ackedPackets = append(ackedPackets, el)
 			}
 		} else {
-			h.ackPacket(el)
-			ackedPackets = append(ackedPackets, congestion.PacketInfo{Number: packetNumber, Length: packet.Length})
+			ackedPackets = append(ackedPackets, el)
 		}
 	}
 
-	if rttUpdated {
-		// Reset counter if a new packet was acked
-		h.consecutiveRTOCount = 0
+	return ackedPackets, nil
+}
+
+func (h *sentPacketHandler) maybeUpdateRTT(largestAcked protocol.PacketNumber, ackDelay time.Duration, rcvTime time.Time) bool {
+	for el := h.packetHistory.Front(); el != nil; el = el.Next() {
+		packet := el.Value
+		if packet.PacketNumber == largestAcked {
+			h.rttStats.UpdateRTT(rcvTime.Sub(packet.SendTime), ackDelay, time.Now())
+			return true
+		}
+		// Packets are sorted by number, so we can stop searching
+		if packet.PacketNumber > largestAcked {
+			break
+		}
 	}
+	return false
+}
 
-	h.garbageCollectSkippedPackets()
+func (h *sentPacketHandler) updateLossDetectionAlarm() {
```

Comment:
Because we don't set the alarm here, only update one particular alarm (h.alarm) that is then set in the session.